Disable and fix tests for when BDB is not compiled

achow101 opened the pull request:
This PR fixes tests for when BDB is not compiled. Tests which rely on or test legacy wallet behavior are disabled and skipped when BDB is not compiled. For the components of some tests that are for legacy wallet things, those parts of the tests are skipped.

For the majority of tests, changes are made so that they can be run with either legacy wallets or descriptor wallets without materially effecting the test. Most tests only need the wallet for balance and transactions, so the type of wallet is not an important part of those tests. Additionally, some tests are wallet agnostic and modified to instead use the test framework's MiniWallet.

achow101 committed:
tests: Don't make any wallets unless wallet is required

achow101 committed:
Disable wallet_descriptor.py bdb format check if BDB is not compiled

achow101 committed:
Disable upgrades tests that require BDB if BDB is not compiled

achow101 committed:
Have feature_filelock.py test both bdb and sqlite, depending on compiled

achow101 committed:
Setup wallets with descriptors for feature_notifications

achow101 committed:
Setup wallets for interface_bitcoin_cli.py

achow101 committed:
tests: Set descriptors default based on compilation

Determines whether descriptors should be used based on whether the
--descriptor or --legacy-wallet option is set,
and the compiled support. If no option is set and both BDB and SQLite
are available, it defaults to legacy.

This is used to switch descriptor agnostic tests between descriptors and
legacy wallet.

achow101 committed:
Use MiniWallet in rpc_net.py

achow101 committed:
Setup wallets for interface_zmq.py

achow101 committed:
Skip legacy wallet reliant tests if BDB is not compiled

achow101 commented on test/functional/test_framework/test_framework.py:
```diff
@@ -767,8 +777,23 @@ def skip_if_no_bitcoind_zmq(self):
+        self.requires_wallet = True
         if not self.is_wallet_compiled():
             raise SkipTest("wallet has not been compiled.")
+        if self.options.descriptors:
+            self.skip_if_no_sqlite()
+        else:
+            self.skip_if_no_bdb()
```
No, `get_use_descriptors()` is used to switch between legacy and descriptors for the tests where the type is not explicitly being set. This is different from the wallet tests (that will call this `skip_if_no_wallet()`) which are testing specific descriptor and legacy behavior.

achow101 commented on test/functional/test_framework/test_framework.py:
```diff
@@ -808,3 +833,23 @@ def is_wallet_tool_compiled(self):
+        Determine whether descriptors should be used based on compilation support and option
+        Should only be used for tests that explicitly check self.is_wallet_compiled()
+        """
+        descriptors = self.options.descriptors
+        if not self.is_bdb_compiled() and self.is_sqlite_compiled() and not descriptors:
+            descriptors = True
+        elif self.is_bdb_compiled() and not self.is_sqlite_compiled() and descriptors:
+            descriptors = False
+        return descriptors
```
Not really.

The purpose of this function is to enable or disable features depending on whether support is compiled and the current settings, for those tests where the descriptors setting is not important to the test and only need to just work.

ysangkok replied:
@mjdietzx , you can use [Quine-McCluskey](https://github.com/ysangkok/py-kmap/blob/master/newqm.py) to derive a minimal function:

```
>>> from newqm import QM
>>> qm = QM(['A','B','C'])
>>> def aj(a,b,c): # a is self.options.descriptors, b is bdb, c is sqlite compilation
...      descriptors = a
...      if not b and c and not a:
...          descriptors = True
...      elif b and not c and a:
...          descriptors = False
...      return descriptors
... 
>>> from itertools import product
>>> values = [idx for (idx,x) in enumerate(product([False,True], repeat=3)) if aj(*x)]
>>> qm.get_function(qm.solve(values,[])[1])
'(((NOT B) AND C) OR (A AND (NOT B)) OR (A AND C))'
```

If we color each conjunctive element separately, you can see which elements of the truth table it covers: https://imgur.com/vOVyQP8

Expanded out using the original bindings:

```
option_set = self.options.descriptors
bdb = self.is_bdb_compiled()
sqlite = self.is_sqlite_compiled()

return ((NOT bdb) AND sqlite) \
  OR (option_set AND (NOT bdb)) \
  OR (option_set AND sqlite)
```

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -1011,9 +1013,11 @@ std::unique_ptr<WalletDatabase> MakeDatabase(const fs::path& path, const Databas
 
     Optional<DatabaseFormat> format;
     if (exists) {
+#ifdef USE_BDB
         if (ExistsBerkeleyDatabase(path)) {
             format = DatabaseFormat::BERKELEY;
         }
+#endif
 #ifdef USE_SQLITE
```
No, it's supposed to `#ifdef ... #ifdef` because both `USE_BDB` and `USE_SQLITE` can both be set, and in fact, is the default.

The instances where `#ifdef...#elif` is used is only because one wallet type must be defaulted to, and for now, we default to BDB.

achow101 commented on test/functional/feature_filelock.py:
```diff
@@ -27,11 +27,20 @@ def run_test(self):
+                    expected_msg = "Error: Error initializing wallet database environment"
+                self.nodes[1].assert_start_raises_init_error(extra_args=['-walletdir={}'.format(wallet_dir), '-wallet=' + self.default_wallet_name, '-noserver'], expected_msg=expected_msg, match=ErrorMatch.PARTIAL_REGEX)
+
+            if self.is_bdb_compiled():
+                check_wallet_filelock(False)
+            if self.is_sqlite_compiled():
+                check_wallet_filelock(True)
```
When both are compiled, we want to test both cases.

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -14,7 +14,9 @@
+#ifdef USE_BDB
 #include <wallet/bdb.h>
+#endif
 #ifdef USE_SQLITE
```
No, if both are compiled, then we need both headers.

achow101 commented on test/functional/wallet_multiwallet.py:
```diff
@@ -101,20 +101,31 @@ def wallet_file(name):
 
         # should raise rpc error if wallet path can't be created
-        assert_raises_rpc_error(-1, "boost::filesystem::create_directory:", self.nodes[0].createwallet, "w8/bad", descriptors=False)
+        err_code = -1
+        if self.options.descriptors:
+            err_code = -4
```
Done in upstream PR #18788

MarcoFalke commented:
```
libbitcoin_wallet.lib(wallet_walletutil.obj) : error LNK2001: unresolved external symbol "bool __cdecl ExistsSQLiteDatabase(class boost::filesystem::path const &)" (?ExistsSQLiteDatabase@@YA_NAEBVpath@filesystem@boost@@@Z) [C:\projects\bitcoin\build_msvc\bench_bitcoin\bench_bitcoin.vcxproj]
```

```
The subject line of commit hash 8ea48f202d800895abea798385ac0e67f7f3034d is followed by a non-empty line. Subject lines should always be followed by a blank line.

^---- failure generated from test/lint/lint-git-commit-check.sh

achow101 commented:
Fixed the linter error.

I'm not sure what is causing the appveyor failure, nor how to fix it.

ryanofsky commented:
> I'm not sure what is causing the appveyor failure, nor how to fix it.

I dug up the old link https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/36097732 and the link errors for unresolved symbols just come from removing wallet/bdb.cpp and wallet/salvage.cpp from original libbitcoin_wallet_a_SOURCES setting in Makefile.am. You might need to add these explicitly to build_msvc/libbitcoin_wallet/libbitcoin_wallet.vcxproj.in. Example of how to do this: 9eaeb7fb8d4ab0d4493849e6c17e314fd75fea9c

ryanofsky commented:
> Example of how to do this: [9eaeb7f](https://github.com/bitcoin/bitcoin/commit/9eaeb7fb8d4ab0d4493849e6c17e314fd75fea9c)

Better simpler example: 0660119ac372c2863d14060ac1bc9bc243771f94

achow101 commented:
> > I'm not sure what is causing the appveyor failure, nor how to fix it.
> 
> I dug up the old link https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/36097732 and the link errors for unresolved symbols just come from removing wallet/bdb.cpp and wallet/salvage.cpp from original libbitcoin_wallet_a_SOURCES setting in Makefile.am. You might need to add these explicitly to build_msvc/libbitcoin_wallet/libbitcoin_wallet.vcxproj.in. Example of how to do this: [9eaeb7f](https://github.com/bitcoin/bitcoin/commit/9eaeb7fb8d4ab0d4493849e6c17e314fd75fea9c)

I think I've fixed it.

ryanofsky reviewed (commented):
Started review (will update list below with progress).

PR title "Disable and fix tests for when BDB is not compiled" does not seem accurate. This goes beyond allowing BDB to be disabled and disabling corresponding tests, but also enables enables new tests and makes fixes not connected to BDB (e.g. "Fix mock SQLiteDatabases" and "Fix change detection of imported internal descriptors.") 

- [X] d4892a3415d8668d6524c2eb843a639c06d317fc Include wallet/bdb.h where it is actually being used (1/20)
- [X] 0cbff9303729dddaad781a4d285c141a86053687 Do not compile BDB things when USE_BDB is defined (2/20)
- [X] cec77056b0945522c0305db8d4754614bfa66009 Enforce salvage is only for BDB wallets (3/20)
- [X] eb2e1680044484833c0e82b133678a77d8660233 RPC: Require descriptors=True for createwallet when BDB is not compiled (4/20)
- [X] 98e8752fa3f8332f73a3e75436523a790364b0ce GUI: Force descriptor wallets when BDB is not compiled (5/20)
- [X] 047d4e3df48f54f1063914cca9e000da33064b01 Allow disabling BDB in configure with --without-bdb (6/20)
- [X] f864402058b4a1a642001bc87d9ef1ce9603f81d Fix mock SQLiteDatabases (7/20)
- [ ] d5157cfa759d2c24c98cda8ae35181c5acbee6b4 Fix change detection of imported internal descriptors (8/20)
- [X] d97e22f5244cb913ef84953309a3eaee07f54c1c Add a helper to determine whether descriptors should be used (9/20)
- [ ] 235bbb7f3f5bba9a72c4e452728ca9ec1b5d5eec Add skip_if_no_bdb and is_bdb_compiled helpers (10/20)
- [ ] 8a0f9fa368236e51b3722ee8f39ba4aca3b78449 Skip legacy wallet reliant tests if BDB is not compiled (11/20)
- [ ] 99f8ffac9c4c16677a9ee39351ba5b5ec299edbb tests: Don't make any wallets unless wallet is required (12/20)
- [ ] 49e9ee040bb4fdb51c0b5ce10d2e5d4cdf1a5c77 Disable wallet_descriptor.py bdb format check if BDB is not compiled (13/20)
- [ ] d53208cf7db7a3d92f4b701cdcdf7c9df0fbee42 Disable upgrades tests that require BDB if BDB is not compiled (14/20)
- [ ] 5fcf045e3299c3e0c6fd7a5d780bf4106c71b334 Have feature_filelock.py test both bdb and sqlite, depending on compiled (15/20)
- [ ] cb126d43dfe3eb324405d9dccf2e5288af7592c7 Setup wallets with descriptors for feature_notifications (16/20)
I- [ ] bd9d5a844ba4150f5d4988104decc46803fe78ad Setup wallets for interface_bitcoin_cli.py (17/20)
- [ ] 02b0ee670a3b4ce450ab1479a2fa56f6080c9c91 Handle when bdb is disabled in rpc_deprecated (18/20)
- [ ] 482ec906b266973d76bd70385621c26136c213eb Use MiniWallet in rpc_net.py (19/20)
- [ ] cb3d26b317556737f3d2652494b4604b22a82ed0 Setup wallets for interface_zmq.py (20/20)

achow101 commented on src/wallet/init.cpp:
```diff
@@ -69,9 +71,12 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
 #endif
     argsman.AddArg("-walletrbf", strprintf("Send transactions with full-RBF opt-in enabled (RPC only, default: %u)", DEFAULT_WALLET_RBF), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
 
+#ifdef USE_BDB
```
Done in upstream #20202

achow101 commented on src/wallet/walletdb.cpp:
```diff
@@ -1059,8 +1063,11 @@ std::unique_ptr<WalletDatabase> MakeDatabase(const fs::path& path, const Databas
+#ifdef USE_BDB
     return MakeBerkeleyDatabase(path, options, status, error);
+#else
+    return nullptr;
```
Done in upstream #20202

achow101 commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -41,6 +41,11 @@ static std::shared_ptr<CWallet> TestLoadWallet(interfaces::Chain& chain)
     DatabaseStatus status;
     bilingual_str error;
     std::vector<bilingual_str> warnings;
+#ifdef USE_BDB
```
Done in upstream #20202

achow101 committed:
Explicitly mark legacy wallet tests as such

Some tests are intended to test only legacy wallet behavior. With
automatic switching of wallet type, we need to make them explicit

achow101 committed:
Require legacy wallet for wallet_upgradewallet.py

achow101 committed:
Fix wallet_send.py wallet setup to work with descriptors

Fixes the wallet setup so this test works with descriptor wallets. Also
enabled explicit descriptor and legacy wallet testing in the test
runner.

achow101 commented on test/functional/test_framework/test_framework.py:
```diff
@@ -824,3 +824,15 @@ def is_zmq_compiled(self):
+        Should only be used for tests that explicitly check self.is_wallet_compiled()
+        """
+        descriptors = self.options.descriptors
+        if not self.is_bdb_compiled() and self.is_sqlite_compiled() and not descriptors:
```
This isn't used to skip tests.

I've made a change that somewhat incorporates what you suggest.

achow101 commented:
Changed to automatically switch to using descriptor wallets if BDB is not compiled and `--legacy-wallet` is not compiled. This requires that several tests have `--legacy-wallet` be added in `test_runner.py`. Additionally, this broke `wallet_send.py` so those tests have to be updated to work with descriptor wallets. `wallet_upgradewallet.py` also had to be restricted to just a legacy wallet test.

luke-jr commented on build_msvc/bitcoin_config.h:
```diff
@@ -38,6 +38,12 @@
+#define USE_BDB 1
+
+/* Define to 1 to enable SQLite wallet */
+#define USE_SQLITE 1
```
Should we be adding this for 0.21?

ryanofsky replied:
> Should we be adding this for 0.21?

No opinion (seems fine, not necessary), but `#define USE_SQLITE 1` is part of base PR #20202. @achow101, PR description could be clearer and say ***This is based on #20202.*** or something along those lines since it seems like I'm not the only person to not realize this PR contains changes from another PR

luke-jr commented on src/wallet/salvage.cpp:
```diff
@@ -27,6 +28,7 @@ bool RecoverDatabaseFile(const fs::path& file_path, bilingual_str& error, std::v
     DatabaseStatus status;
     options.require_existing = true;
     options.verify = false;
+    options.require_format = DatabaseFormat::BERKELEY;
```
Is this another bugfix we should put in 0.21?

achow101 replied:
It's not strictly necessary. Other stuff later like trying to actually open the wallet file if it isn't BDB will still fail and we won't do anything to the file. It will just have a slightly worse error message.

ysangkok commented on doc/build-unix.md:
```diff
@@ -41,7 +41,7 @@ Optional dependencies:
  ------------|------------------|----------------------
  miniupnpc   | UPnP Support     | Firewall-jumping support
- libdb4.8    | Berkeley DB      | Wallet storage (only needed when wallet enabled)
+ libdb4.8    | Berkeley DB      | Optional, wallet storage (only needed when wallet enabled)
```
the sentence makes it sound like it is optional because the wallet can be disabled. wouldn't it be better to say "Optional. Needed to open legacy wallets."

luke-jr replied:
Probably shouldn't use the word "legacy" (at least user-facing) while it isn't actually legacy yet.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/20267#issue-512655353

> Fixes the tests for #20202. When BDB is not compiled, the wallet tests that rely on the legacy wallet will be disabled. Other tests not necessarily requiring the wallet, but can optionally test behavior with the wallet, are updated to handle the different wallet types. One test, `rpc_net.py`, which only uses the wallet to setup a few things, is changed to use the test framework's MiniWallet instead.
> 
> This PR is separate from and based on #20202 because it needs a few other PRs to make the tests work: #18788, #20262, and #20266

Starting to look at this but can the description be updated to maybe strip out the references to old PRs and say what behavior change this PR is implementing? What happens with test runner currently, and what happens after this PR? What happens running individual BDB tests currently, and what happens after?

I'm used to test runner showing a table of all tests `Passed` or `Failed` or `Skipped`. Here, it seems like in the absence of BDB some tests are partially skipped so I would expect to see some new `Passed (partial)` status instead of plain `Passed`. Or to avoid this, I'd expect these tests to be split and run twice with `--legacy-wallet` and `--descriptors`, so there is never any partially passed test only fully passed or fully skipped.

I guess I'm just confused about what the exact goal is for this PR, so it would be good to spell out  the new behavior.

I'm also confused about why some code is looking at self.options.descriptors, and other code is calling `self.get_use_descriptors()`. And `self.get_use_descriptors()` is using `self.options.descriptors` internally but there is also `self.options.descriptors = self.get_use_descriptors()` so some kind of caching or recursive dependency? This might be worth simplifying or explaining some place.

achow101 commented:
> Starting to look at this but can the description be updated to maybe strip out the references to old PRs and say what behavior change this PR is implementing? What happens with test runner currently, and what happens after this PR? What happens running individual BDB tests currently, and what happens after?

Updated the OP.

> I'm used to test runner showing a table of all tests `Passed` or `Failed` or `Skipped`. Here, it seems like in the absence of BDB some tests are partially skipped so I would expect to see some new `Passed (partial)` status instead of plain `Passed`. Or to avoid this, I'd expect these tests to be split and run twice with `--legacy-wallet` and `--descriptors`, so there is never any partially passed test only fully passed or fully skipped.

We already have tests which are "partial" as some parts are not tested if the wallet is not compiled. If we do want to have a "partial" indicator for tests where we disable components based on compiled support, then I think it should be done in a followup.

> I'm also confused about why some code is looking at self.options.descriptors, and other code is calling `self.get_use_descriptors()`. And `self.get_use_descriptors()` is using `self.options.descriptors` internally but there is also `self.options.descriptors = self.get_use_descriptors()` so some kind of caching or recursive dependency? This might be worth simplifying or explaining some place.

This is a holdover from a previous iteration. I've removed the extra uses of `self.get_use_descriptors()` and just use `self.options.descriptors`. The purpose and behavior of `self.get_use_descriptors` is described in the OP.

laanwj commented:
Concept ACK

ryanofsky commented:
Thanks for updates. Init sequence still seems too complicated. I don't see a reason options can't be set once simply in the constructor. In the current implementation cf5de8ec59ac6e2a49c728b886eb5d46d9705d1e:

- self.options.descriptors value changes over time and is set to the return value of a function which circularly reads self.options.descriptors
- default_wallet_name can't be read/written from setup_params and is a dynamic property executing code and confusingly depending on the changing self.options.descriptors value

The following cleanups seem to work all right:

```diff
diff --git a/test/functional/test_framework/test_framework.py b/test/functional/test_framework/test_framework.py
index c0289bf491e..76b3815d25b 100755
--- a/test/functional/test_framework/test_framework.py
+++ b/test/functional/test_framework/test_framework.py
@@ -102,6 +102,7 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
         self.supports_cli = True
         self.bind_to_localhost_only = True
         self.parse_args()
+        self.default_wallet_name = "default_wallet" if self.options.descriptors else ""
         self.wallet_data_filename = "wallet.dat"
         # Optional list of wallet names that can be set in set_test_params to
         # create and import keys to. If unset, default is len(nodes) *
@@ -118,10 +119,6 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
             self.options.timeout_factor = 99999
         self.rpc_timeout = int(self.rpc_timeout * self.options.timeout_factor) # optionally, increase timeout by a factor
 
-    @property
-    def default_wallet_name(self):
-        return "default_wallet" if self.options.descriptors else ""
-
     def main(self):
         """Main function. This should not be overridden by the subclass test scripts."""
 
@@ -201,6 +198,18 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
         self.options = parser.parse_args()
         self.options.previous_releases_path = previous_releases_path
 
+        config = configparser.ConfigParser()
+        config.read_file(open(self.options.configfile))
+        self.config = config
+
+        if self.options.descriptors is None:
+            # Prefer BDB unless it isn't available
+            if self.is_bdb_compiled():
+                self.options.descriptors = False
+            elif self.is_sqlite_compiled():
+                self.options.descriptors = True
+            # If neither are compiled, tests requiring a wallet will be skipped and the value of self.options.descriptors won't matter
+
     def setup(self):
         """Call this method to start up the test framework object with options set."""
 
@@ -210,9 +219,8 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
 
         self.options.cachedir = os.path.abspath(self.options.cachedir)
 
-        config = configparser.ConfigParser()
-        config.read_file(open(self.options.configfile))
-        self.config = config
+        config = self.config
+
         fname_bitcoind = os.path.join(
             config["environment"]["BUILDDIR"],
             "src",
@@ -231,9 +239,6 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
             os.path.join(config['environment']['BUILDDIR'], 'src', 'qt'), os.environ['PATH']
         ])
 
-        # Determine descriptors default
-        self.options.descriptors = self.get_use_descriptors()
-
         # Set up temp directory and start logging
         if self.options.tmpdir:
             self.options.tmpdir = os.path.abspath(self.options.tmpdir)
@@ -847,14 +852,3 @@ class BitcoinTestFramework(metaclass=BitcoinTestMetaClass):
     def is_bdb_compiled(self):
         """Checks whether the wallet module was compiled with BDB support."""
         return self.config["components"].getboolean("USE_BDB")
-
-    def get_use_descriptors(self):
-        descriptors = self.options.descriptors
-        if descriptors is None:
-            # Prefer BDB unless it isn't available
-            if self.is_bdb_compiled():
-                descriptors = False
-            elif self.is_sqlite_compiled():
-                descriptors = True
-            # If neither are compiled, tests requiring a wallet will be skipped and the value of self.options.descriptors won't matter
-        return descriptors
```

Would be great to see these changes backported so this PR is easy to understand and review without  @<!-- -->property and get_use_descriptors complications

achow101 committed:
tests: Disable bdb dump test when no bdb

achow101 commented:
Included @ryanofsky's suggestions. Also rebased to fix two silent merge conflicts.

ryanofsky commented on test/functional/wallet_send.py:
```diff
@@ -168,49 +169,91 @@ def run_test(self):
-        self.nodes[1].createwallet(wallet_name="w2")
+        self.nodes[1].createwallet(wallet_name="w2", blank=True)
         w2 = self.nodes[1].get_wallet_rpc("w2")
+        xpriv = "tprv8ZgxMBicQKsPfHCsTwkiM1KT56RXbGGTqvc2hgqzycpwbHqqpcajQeMRZoBD35kW4RtyCemu6j34Ku5DEspmgjKdt2qe4SvRch5Kk8B8A2v"
```
In commit "Fix wallet_send.py wallet setup to work with descriptors" (1fa9bbdb4076944272d5246d1140d29ba3ecb3a3)

This seems like the same code earlier copied from the feature_notifications test. You can see earlier comment, but I think it would be better to have a framework or util function that can easily create wallets with the same seed. (It would also seem better not to hard code that seed, but the hardcoding is less a problem than the complexity and boilerplate these changes add to the tests)

achow101 replied:
Done

ryanofsky reviewed (approved):
Code review ACK 5a985cd6656afe976f73b3276017e22055a8050c. I think this PR is a nice improvement, getting tests less tied to the legacy wallet and also making test init order more rational and making behavior more explicit. I did leave a lot of suggestions for improvement and think it would be great if you could look into them.

---

re: Summary https://github.com/bitcoin/bitcoin/pull/20267#issue-512655353

> Fixes the tests for #20202.

Ideally the PR would have a summary that doesn't require digging into other issues, and indicate up front why would reviewers might be interested in this PR. Would change this to something like: "This PR fixes tests that currently fail when bitcoin is configured with the `--without-bdb` option, which was recently added in #20202."

> When BDB is not compiled, the wallet tests that rely on the legacy wallet will be disabled. Other tests not necessarily requiring the wallet, but can optionally test behavior with the wallet, are updated to handle the different wallet types. One test, `rpc_net.py`, which only uses the wallet to setup a few things, is changed to use the test framework's MiniWallet instead.
>
> For the majority of tests, changes are made so that they can be run with either legacy wallets or descriptor wallets without materially effecting the test. Most tests only need the wallet for balance and transactions, so the type of wallet is not an important part of those tests.
>
> The tests which rely on or test specific legacy wallet behavior are disabled or skipped if BDB is not compiled. Since these tests can be testing for general wallet things as well, the test file itself may not always be outright skipped, but rather some components of the tests skipped.

This is very jumbled. It talks about disabled tests two places (at the beginning and end) and buries the most helpful section about the majority of tests in the middle. It also oddly describes the majority of tests strategy after describing one special case of it.

I think a better way to structure this would be. "For the most part, this PR tries to update tests which depend on legacy behavior, to work with both legacy and descriptor wallets. In one case, rpc_net.py this was done by [...]. In cases where tests are more tied to legacy features, whole tests or portions of tests are skipped."

> Lastly a helper function `get_use_descriptors()` is added which determines whether descriptors should be used. It returns whatever was given in the command line. If no command line option was given, it will default to BDB if it as compiled, otherwise it will use SQLite (descriptors). If neither are compiled, then the wallet should be disabled and the option won't be used. `get_use_descriptors` is used to set the value for `self.options.descriptors` when neither `--legacy-wallet` nor `--descriptors` is given.

This would be a helpful comment to have in the code on the `get_use_descriptors` function but I think it is out of place here and better to drop from the PR description.

achow101 commented on test/functional/wallet_descriptor.py:
```diff
@@ -23,11 +23,12 @@ def skip_test_if_missing_module(self):
-        wallet_info = self.nodes[0].getwalletinfo()
-        assert_equal(wallet_info['format'], 'bdb')
-        self.nodes[0].unloadwallet("legacy1")
+        if self.is_bdb_compiled():
```
I think that the optimal solution for these is to split them into separate tests for wallet and non-wallet components.

achow101 commented on test/functional/feature_filelock.py:
```diff
@@ -27,11 +29,21 @@ def run_test(self):
+                    expected_msg = "Error: Error initializing wallet database environment"
+                self.nodes[1].assert_start_raises_init_error(extra_args=['-walletdir={}'.format(wallet_dir), '-wallet=' + wallet_name, '-noserver'], expected_msg=expected_msg, match=ErrorMatch.PARTIAL_REGEX)
+
+            if self.is_bdb_compiled():
+                check_wallet_filelock(False)
+            if self.is_sqlite_compiled():
+                check_wallet_filelock(True)
```
#20892 runs such tests twice. It also seems like this should be one of those tests that the wallet portion is split into a separate test.

achow101 commented on test/functional/feature_notifications.py:
```diff
@@ -49,6 +49,30 @@ def setup_network(self):
     def run_test(self):
+        if self.is_wallet_compiled():
+            # Setup the descriptors to be imported to the wallet
+            seed = "cTdGmKFWpbvpKQ7ejrdzqYT2hhjyb3GPHnLAK7wdi5Em67YLwSm9"
+            xpriv = "tprv8ZgxMBicQKsPfHCsTwkiM1KT56RXbGGTqvc2hgqzycpwbHqqpcajQeMRZoBD35kW4RtyCemu6j34Ku5DEspmgjKdt2qe4SvRch5Kk8B8A2v"
```
Until we implement private descriptor export, that won't be possible.

achow101 commented on test/functional/test_framework/test_framework.py:
```diff
@@ -186,15 +186,27 @@ def parse_args(self):
         self.options = parser.parse_args()
         self.options.previous_releases_path = previous_releases_path
 
+        config = configparser.ConfigParser()
+        config.read_file(open(self.options.configfile))
+        self.config = config
```
The `is_*_compiled` checks require self.config to already exist, so it must come before them and can't be at the bottom of the function.

achow101 commented on test/functional/test_framework/test_framework.py:
```diff
@@ -186,15 +186,27 @@ def parse_args(self):
+                self.options.descriptors = False
+            elif self.is_sqlite_compiled():
+                self.options.descriptors = True
+            # If neither are compiled, tests requiring a wallet will be skipped and the value of self.options.descriptors won't matter
```
Done

achow101 replied:
Reverted this since it is causing problems when configured with `--disable-wallet`

achow101 commented on test/functional/test_framework/test_framework.py:
```diff
@@ -411,6 +414,8 @@ def import_deterministic_coinbase_privkeys(self):
             self.init_wallet(i)
 
     def init_wallet(self, i):
+        if not self.requires_wallet:
```
Done.

achow101 commented on test/functional/feature_notifications.py:
```diff
@@ -49,6 +49,30 @@ def setup_network(self):
 
     def run_test(self):
+        if self.is_wallet_compiled():
+            # Setup the descriptors to be imported to the wallet
```
Done

achow101 commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -63,6 +63,8 @@ def run_test(self):
         self.log.info("Test -getinfo returns expected network and blockchain info")
         if self.is_wallet_compiled():
+            self.nodes[0].createwallet(wallet_name=self.default_wallet_name, descriptors=self.options.descriptors)
+            self.nodes[0].importprivkey(privkey=self.nodes[0].get_deterministic_priv_key().key, label='coinbase')
```
Done

achow101 commented on test/functional/rpc_net.py:
```diff
@@ -48,6 +49,8 @@ def set_test_params(self):
         self.supports_cli = False
 
     def run_test(self):
+        self.wallet = MiniWallet(self.nodes[0])
```
Added a comment. It is only used to generate a transaction so that `last_transaction` in `getpeerinfo` can be checked.

achow101 commented on test/functional/interface_zmq.py:
```diff
@@ -68,6 +68,12 @@ def skip_test_if_missing_module(self):
+            self.nodes[0].createwallet(wallet_name=self.default_wallet_name, descriptors=self.options.descriptors, load_on_startup=True)
+            self.nodes[0].importprivkey(privkey=self.nodes[0].get_deterministic_priv_key().key, label='coinbase')
+            self.nodes[1].createwallet(wallet_name=self.default_wallet_name, descriptors=self.options.descriptors, load_on_startup=True)
+            self.nodes[1].importprivkey(privkey=self.nodes[1].get_deterministic_priv_key().key, label='coinbase')
```
Yes. Done

achow101 commented on test/functional/mempool_compatibility.py:
```diff
@@ -23,6 +23,7 @@ class MempoolCompatibilityTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 2
         self.wallet_names = [None]
+        self.requires_wallet = True
```
Perhaps, but that's a bit more complicated.

achow101 commented on test/functional/wallet_send.py:
```diff
@@ -241,11 +284,13 @@ def run_test(self):
-        res = w2.walletprocesspsbt(res["psbt"])
-        assert res["complete"]
+        if not self.options.descriptors:
+            # This tests legacy watch-only behavior only.
```
Added a comment.

achow101 commented on test/functional/wallet_send.py:
```diff
@@ -168,49 +169,91 @@ def run_test(self):
 
+        if not self.options.descriptors:
+            # w4 has private keys enabled, but only contains watch-only keys (from w2)
+            # This is legacy wallet behavior only
```
Done

achow101 commented:
Rebased. I think there are improvements that can be done in followup PRs. For example, #20892 implements running a test twice. But for now, I would like for this to be merged so that we can get more test coverage of descriptor wallets.

laanwj commented:
CI fail looks like an actual issue:
```
.........Traceback (most recent call last):
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/create_cache.py", line 27, in <module>
    CreateCache().main()
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 105, in __init__
    self.default_wallet_name = "default_wallet" if self.options.descriptors else ""
AttributeError: 'Namespace' object has no attribute 'descriptors'
```

achow101 commented:
> CI fail looks like an actual issue:

Should be resolved now.

ryanofsky reviewed (approved):
Code review ACK a831ba6014827d5afc9b423773361350359c94d5. Changes since last review: implementing many review suggestions (thanks!) adding comments, tweaking logic in small ways, and simplifying the PR description. Also dropped rpc_deprecated.py commit 42b627a3a3ab8ee30db5fa3f6b14be5869c2fcfe no longer needed after #20891

Would encourage others to review this. There are a lot of commits but almost all the commits are trivial (make sure to tick the ignore whitespace option).

Two things it would be good to clean up later:

1. It would be good to remove duplicate code added to wallet_send and feature_notifications tests (https://github.com/bitcoin/bitcoin/pull/20267#discussion_r556927365 and https://github.com/bitcoin/bitcoin/pull/20267#discussion_r556977983) that creates different descriptor wallets with the same seeds. From comments it sounds like when more descriptor import stuff is implemented this code can be simplified by reusing existing seeds instead of hardcoding new seeds, but it seems like the tests could be more readable and maintainable even before this if they were refactored to call a common helper function to set up the wallets, instead of adding extra setup to already complicated tests.
2. IMO, handling of sqlite/bdb test variants is moving in a bad direction here and in followup #20892. I think individual test invocations should be as small and fast as possible, and broken up much as possible. I think invoking a test should run one single variant and test runner should responsible for iterating over variants. Advantages of this approach:
   * Faster debug cycles. Tests that fail faster can be debugged faster. Broken up tests can also be parallelized more.
   * More accurate reporting. There's an increasing number of tests that are skipping checks based on compile options and misleadingly reporting skipped checks as passing. We could fix this by expanding test result status from Passing / Failed / Skipped to include a new _Passing but partially skipped_ status. But it would be clearer to avoid this and to break these tests up so "Passing" can go back to meaning fully passing with no checks skipped.
   * More straightforward usage. A few tests are ignoring descriptor/legacy options which is inconsistent and confusing.
   * More readable code. Smaller tests are easier to understand and debug, and some tests are doing things like repeating code or defining unnecessary functions and calling them twice to cover multiple variants. Tests would be simpler if test_runner managed all the variations, and tests could just obey options instead of listing variants.

S3RK commented on test/functional/mempool_compatibility.py:
```diff
@@ -23,6 +23,7 @@ class MempoolCompatibilityTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 2
         self.wallet_names = [None]
+        self.requires_wallet = True
```
I can't understand why do we need "the old versions have their wallets' setup". I've tried to run this test with the flag set to False and it makes no difference

achow101 replied:
It's redundant now. Previously this variable was checked inside of `init_wallet()` which was called by `import_deterministic_coinbase_privkeys()` so we needed to set it to `True` to allow our call of that to work. But since checking `requires_wallet` is moved outside of that, it is no longer needed.

Removed and dropped the commit.

S3RK commented:
Started doing review. Have only one small question, but I'll get back with more detailed feedback later

ryanofsky reviewed (approved):
Code review ACK 49797c3ccfbb9f7ac9c1fbb574d35b315c103805. Only change since last review is dropping last commit. Previous review w/ suggestions for future followup is https://github.com/bitcoin/bitcoin/pull/20267#pullrequestreview-581508843

laanwj commented:
> IMO, handling of sqlite/bdb test variants is moving in a bad direction here and in followup #20892. I think individual test invocations should be as small and fast as possible, 

Yes, I think so too. And agree it's good if skipping tests is reported explicitly. However I don't think this PR makes this much worse. But yes it does add some `if self.is_wallet_compiled():`.

ACK 49797c3ccfbb9f7ac9c1fbb574d35b315c103805

laanwj commented:
UBsan error is:
```
test/fuzz/system.cpp:57:159: runtime error: implicit conversion from type 'int' of value -2049 (32-bit, signed) to type 'unsigned int' changed the value to 4294965247 (32-bit, unsigned)
```
during the fuzz tests. As these are not touched here, I don't think it needs to block merge.
